feat: wg-easy add preflights

nvanthao opened the pull request:
# Add Preflight Checks to WireGuard Easy (wg-easy)

## Overview
This PR adds preflight checks to the `wg-easy` chart.

## Implementation Details
- Each chart-specific preflight is defined as a Kubernetes Secret with label `troubleshoot.sh/kind: preflight`
- Preflight specs are stored in each chart's templates directory:

  ```bash
  cert-manager/templates/secret-preflights.yaml
  wg-easy/templates/secret-preflights.yaml
  ```
- Added new task `helm-preflight` to streamline verification
- Updated container image to include the `preflight` binary

## Current Status
- The PR is based on https://github.com/replicatedhq/platform-examples/pull/55 and will be rebased once that PR is merged
- The `cert-manager` preflight is currently a placeholder to demonstrate multiple preflight specs merging, not the final preflight content
- The `wg-easy` preflight has been implemented with actual checks relevant to the chart

## Testing
Locally verify preflight specs with:

```bash
helm template <chart-dir> | preflight - --dry-run
```

## Next Steps
- [ ] Finalize cert-manager preflights with actual checks
- [ ] Consider adding preflights to additional charts

chris-sanders reviewed (changes requested):
I was trying to do a review and then I realized you've based this on another active development branch but targeted the PR against main.

The problem is that means I'm not reviewing both PR's and I can't tell what you've changed and what the other PR is chaning.

If you base a PR on top of another existing branch I think we need to review against that other PR until it's merged so that the review is only of your current changes not everything from both branches.

chris-sanders reviewed (changes requested):
Just watching the asciinema and again I don't know for sure if this is feedback from your PR or the branch you're building on top of but I'll note what I see either way.

1. You ran a command 'ag' which I'm guessing is an alias or shortcut you've setup yourself? If so you shouldn't do that in a demo I don't know what that command does, hence I can't understand what you're trying to show us by running it.
2. You've exposed your REPLICATED_API_TOKEN please rotate it and be more careful in the future about recording sensitive values.
3. The output here is far to verbose, that's part of the cause of item 2 above, slim it down we shouldn't be showing task commands like you show here. Especially when they have API TOKENS in them. That goes for both the green and the white text, it's noisy look at the Embedded Cluster CLI outputs for a target of what good looks like. I think just disable all of the verbose output would be a good start.
4. As I mentioned in yesterdays' standup, you shouldn't need cluster-create and setup-kubeconfig in your dependencies. As far as I can tell just setup-kubeconfig is sufficient because it already includes creating a cluster. We need to be careful not to unnecessarily re-run and steps for no reason.
5. This is the biggest one, you're using a bash loop to make this work which is fine for development purposes however how do you see this working in a real install? An end user won't be using Taskfile or this repository, the preflight needs to work through Enterprise Portal and Embedded Cluster installs. Are you addressing that next? I'm fine with this implementation for developers but before we land anything we need to know what this means for end users.

nvanthao commented:
many thanks @chris-sanders 

1. I use [ag](https://github.com/ggreer/the_silver_searcher) as a drop-in replacement for `grep` as it is faster
2. many thanks I have rotated the token for the demo
3. I removed `-v=5` flag
4. I removed dep `cluster-create`
5. In real install, let me do a quick record of what I am seeing and share the thoughts with us in a Loom

nvanthao commented:
>This is the biggest one, you're using a bash loop to make this work which is fine for development purposes however how do you see this working in a real install? An end user won't be using Taskfile or this repository, the preflight needs to work through Enterprise Portal and Embedded Cluster installs. Are you addressing that next? I'm fine with this implementation for developers but before we land anything we need to know what this means for end users.

I've shared the Loom in our Slack channel, tldr;

- for KOTS install in existing cluster, the preflights will run during kots install
- for EC install, the preflight will happen in KOTS Admin Console UI
- for helm install, Vendor Portal contains the instructions to run helm template | preflight per chart
- for helm install, Enterprise Portal is missing preflights run instructions as of now

chris-sanders commented on applications/wg-easy/taskfiles/container.yml:
```diff
@@ -1,90 +1,111 @@
+        ]
     cmds:
+      - task: check-image-exists
+      # Start with host networking for kubectl port-forward compatibility
```
The security implications of this are a bit concerning, although if it's necessary it's not something I would loose sleep over.

Can you explain why kubectl port-forward is needed? Is this running port-forward from within the container, or external to the container? Is there a reason you can't just expose a port rather than use host networking?

chris-sanders reviewed (changes requested):
I just realized this is still targeting main and not the other PR.
I'm assuming all changes in this PR are yours since you asked for another review, is that accurate?

nvanthao commented:
> I just realized this is still targeting main and not the other PR. I'm assuming all changes in this PR are yours since you asked for another review, is that accurate?

Hi @chris-sanders, I can't target the other PR branch as the branch is in a forked repo. As mentioned earlier, I will wait for that PR to be merged to main and rebase this PR.

I've updated the preflights for cert-manager accordingly. I can only find the k8s version as the only prerequisite [here](https://github.com/cert-manager/cert-manager/blob/master/deploy/charts/cert-manager/README.template.md#prerequisites)

chris-sanders commented:
I didn't realize the other PR was in a forked repo, I'll mention to the team to work out of this repo in the future.

chris-sanders commented on applications/wg-easy/charts/cert-manager/templates/secret-preflights.yaml:
```diff
@@ -0,0 +1,28 @@
+              - fail:
+                  when: "< 1.22.0"
+                  message: The application requires at least Kubernetes 1.22.0, and recommends 1.25.0.
+                  uri: https://kubernetes.io
```
What is this uri reference? I don't feel like a generic reference to kubernetes.io is very helpful but I'm not clear what this is for.

Wouldn't a reference to the cert-manager docs bet more appropriate here?

nvanthao replied:
Thank you! I've updated the uri to https://cert-manager.io/docs/installation/helm

chris-sanders commented on applications/wg-easy/Taskfile.yaml:
```diff
@@ -172,6 +172,36 @@ tasks:
     deps:
       - cluster-create
 
+  helm-preflight-dryrun:
```
Do we really need a separate command just for a dry-run? This can't be combined with the helm-preflight command with an argument of some sort?

nvanthao replied:
Great idea, I've removed this task and update `helm-preflight` with `DRY_RUN` argument, we can now run

```
task helm-preflight
task helm-preflight DRY_RUN=true
```

chris-sanders commented on applications/wg-easy/Taskfile.yaml:
```diff
@@ -172,6 +172,36 @@ tasks:
+  helm-preflight-dryrun:
+    desc: Verify the preflight specs that will be run
+    silent: false
+    cmds: 
+      - |
+        TEMP_FILE=$(mktemp)
+        
+        # Find all charts and append their templates to the temp file
+        for chart_dir in $(find charts/ -maxdepth 1 -mindepth 1 -type d); do
+          helm template "$chart_dir" >> "$TEMP_FILE"
+          echo "---" >> "$TEMP_FILE" # Add separator between charts
+        done
+        
+        # Run preflight once on the combined templates
+        echo "Running preflight checks on all templates"
+        cat "$TEMP_FILE" | kubectl preflight - --dry-run
+        
+        # Clean up
+        rm "$TEMP_FILE"
```
Help me understand the value that this provides. 
What is the output of running this when:
1. There's a failure
2. It's successful

Is this meant to check formatting of the preflights?
If something is formatted poorly won't it be hard to debug since you concat them all into a temporary file? It seems like running this on each chart individually would be much more straight forward since you would know which chart failed and the line numbers on errors (does it provide those?) would line up.

nvanthao replied:
Great catch on the failure, I separate the preflight call now to detect if there's failure.

chris-sanders commented on applications/wg-easy/Taskfile.yaml:
```diff
@@ -172,6 +172,36 @@ tasks:
+    desc: Run preflight checks on Helm charts using preflight CLI
+    silent: false
+    cmds: 
+      - defer: rm -f preflightbundle-*.tar.gz
```
The fact that Taskfile can 'defer' is pretty nice I didn't know that before your changes.

However, isn't removing the preflight counter productive? If a preflight fails the user may want to look at the failure or provide it to support for help. At the very least shouldn't there be a way to not remove it so that it's available?

I think my personal preference would be to either:
1. Remove preflightbundle-*.tar.gz before we run the preflight so that only the most recent bundle is present.
2. Don't even try to remove them, just make sure it's in .gitignore and let people clean them up as they see fit.

nvanthao replied:
Gotcha, I opted in for 2 and updated .gitignore

chris-sanders reviewed (changes requested):
Can you be sure to update the docs as well, that would also probably answer some of the questions I had in the review.

nvanthao commented on applications/wg-easy/charts/cert-manager/templates/secret-preflights.yaml:
```diff
@@ -0,0 +1,28 @@
+                  message: Your cluster meets the minimum version of Kubernetes, but we recommend you update to 1.25.0 or later.
+                  uri: https://cert-manager.io/docs/installation/helm/#prerequisites
+              - pass:
+                  message: Your cluster meets the recommended and required versions of Kubernetes.
```
I think that is great improvement using both `Files.Get` or `include`. I opted in for `include` now as from Helm [docs](https://helm.sh/docs/chart_template_guide/accessing_files/)

>Some files cannot be accessed through the .Files object, usually for security reasons.
Files in templates/ cannot be accessed.
Files excluded using .helmignore cannot be accessed.

We will have to create new folder on same level with `templates` which would unnecessarily complicate our directory structure. Also `include` allows future preflight checks to access Helm values more easily.

nvanthao commented on applications/wg-easy/Taskfile.yaml:
```diff
@@ -172,6 +172,25 @@ tasks:
+  helm-preflight:
+    desc: Run preflight checks on Helm charts using preflight CLI (use DRY_RUN=true for dry-run)
+    silent: false
+    vars:
+      DRY_RUN: '{{.DRY_RUN | default "false"}}'
```
yes I think the variable is already scoped locally to the `helm-preflight` task only.

I committed the suggestion but it looks like it is not valid Taskfile syntax

```
err:  "default" is not a valid variable type. Try "sh", "ref", "map" or using a scalar value
file: /Users/gerard/dev/platform-examples/applications/wg-easy/Taskfile.yaml:180:9
  178 |     vars:
  179 |       DRY_RUN:
> 180 |         default: false
      |         ^
  181 |     cmds: 
  182 |       - |
```

chris-sanders replied:
Ok so the syntax wasn't correct but it can still be done can't it?

```yaml
vars:
  DRY_RUN: false
```

Or something like that? 

I'm asking why you're explicitly making a dynamic variable when you don't need it to be dynamic. I think you're copying a pattern that we setup when we were setting global variables and importing environments but I don't think it is necessary here.

Maybe I'm just misunderstanding Taskfile and this isn't possible or the dynamic nature of using "{{.DRY_RUN}}" provides some value that I'm not aware of. Does including "{{.DRY_RUN}}" provide some value or is that the only way to set a variable?

chris-sanders reviewed (commented):
Just looking for clarity on one last thing the rest looks good to me.

nvanthao commented:
The dynamic variable is so that user can toggle between validation mode and execution mode of preflight without modifying the Taskfile.

- `task helm-preflight DRY_RUN=true` - runs in validation mode, checking content of preflight spec to run without executing anything
- `task helm-preflight` - runs the full preflight checks against the cluster.

chris-sanders reviewed (approved):
lgtm
